src/test/scala/chisel3/TypeEquivalenceSpec.scala

```diff
@@ -0,0 +1,268 @@
+  }
+
+  it should "support Vecs" in {
+    val x = Vec(2, UInt(8.W))
+    val y = Vec(2, UInt(8.W))
+    val z = Vec(3, UInt(8.W))
+    assert(x.typeEquivalent(y))
+    assert(!x.typeEquivalent(z))
+  }
+
+  it should "reject nested width mismatches" in {
+    val x = new Bar(8)
+    val y = new Bar(4)
+    assert(!x.typeEquivalent(y))
+
+    val a = Vec(4, new Bar(8))
+    val b = Vec(4, new Bar(4))
+    assert(!a.typeEquivalent(b))
+
+    val c = new Bundle {
+      val foo = new Bar(8)
+    }
+    val d = new Bundle {
+      val foo = new Bar(4)
+    }
+    assert(!c.typeEquivalent(d))
+  }
+
+  it should "support equivalent ChiselEnums" in {
+    val x = Fizz()
+    val y = Fizz()
+    val z = Buzz()
+    assert(x.typeEquivalent(y))
+    assert(!x.typeEquivalent(z))
+  }
+
+  it should "support comparing Analogs" in {
+    val x = Analog(8.W)
+    val y = Analog(8.W)
+    val z = Analog(4.W)
+    assert(x.typeEquivalent(y))
+    assert(!x.typeEquivalent(z))
+  }
+
+  it should "support DontCare" in {
+    assert(DontCare.typeEquivalent(DontCare))
+    assert(!DontCare.typeEquivalent(UInt()))
+  }
+
+  it should "support AsyncReset" in {
+    assert(AsyncReset().typeEquivalent(AsyncReset()))
+    assert(!AsyncReset().typeEquivalent(Bool()))
+    assert(!AsyncReset().typeEquivalent(Clock()))
+    assert(!AsyncReset().typeEquivalent(Reset()))
+  }
+
+  it should "support Clock" in {
+    assert(Clock().typeEquivalent(Clock()))
+  }
+
+  it should "support abstract Reset" in {
+    assert(Reset().typeEquivalent(Reset()))
+  }
+
+  behavior.of("Data.findFirstTypeMismatch")
+
+  it should "support comparing UInts" in {
+    val x = UInt(8.W)
+    val y = UInt(8.W)
+    val z = UInt(4.W)
+    val zz = UInt()
+    x.findFirstTypeMismatch(y, true, true) should be(None)
+    x.findFirstTypeMismatch(z, true, true) should be(
+      Some(": Left (UInt<8>) and Right (UInt<4>) have different widths.")
+    )
+    x.findFirstTypeMismatch(z, true, false) should be(None)
+    x.findFirstTypeMismatch(zz, true, true) should be(Some(": Left (UInt<8>) and Right (UInt) have different widths."))
+    x.findFirstTypeMismatch(zz, true, false) should be(None)
+  }
+
+  it should "support comparing SInts" in {
+    val x = SInt(8.W)
+    val y = SInt(8.W)
+    val z = SInt(4.W)
+    val zz = SInt()
+    x.findFirstTypeMismatch(y, true, true) should be(None)
+    x.findFirstTypeMismatch(z, true, true) should be(
+      Some(": Left (SInt<8>) and Right (SInt<4>) have different widths.")
+    )
+    x.findFirstTypeMismatch(z, true, false) should be(None)
+    x.findFirstTypeMismatch(zz, true, true) should be(Some(": Left (SInt<8>) and Right (SInt) have different widths."))
+    x.findFirstTypeMismatch(zz, true, false) should be(None)
+  }
+
+  it should "catch comparing SInts and UInts" in {
+    val x = UInt(8.W)
+    val y = SInt(8.W)
+    x.findFirstTypeMismatch(y, true, true) should be(Some(": Left (UInt<8>) and Right (SInt<8>) have different types."))
```

Comment:
Curious why the toStrings of left and right are not providing the full path? I feel like that could be helpful in certain situations

Reply:
If you see the later tests, it includes path information for aggregates before the `:`, the idea is that this will be integrated into other checks. I think it may need some massaging but the basic path information is there. It's a little awkward for Vecs since it applies to all elements so I just signify it with `[_]`. I am open to suggestions, perhaps we can refine it on the follow on PR which actually uses the new API.